pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -910,18 +914,40 @@ func GetSelectorRegexMap(l7 policy.L7DataMap) (CachedSelectorREEntry, error) {
 		reStrings := make([]string, 0, len(l7Rules.DNS))
 		for _, dnsRule := range l7Rules.DNS {
 			if len(dnsRule.MatchName) > 0 {
-				dnsRuleName := strings.ToLower(dns.Fqdn(dnsRule.MatchName))
-				dnsPatternAsRE := matchpattern.ToRegexp(dnsRuleName)
-				reStrings = append(reStrings, "("+dnsPatternAsRE+")")
+				dnsRuleName := FromFqdn(dnsRule.MatchName)
+				dnsPatternAsRE, _ := CustomToRegexp(Reverse(dnsRuleName))
+				reStrings = append(reStrings, dnsPatternAsRE)
 			}
 			if len(dnsRule.MatchPattern) > 0 {
-				dnsPattern := matchpattern.Sanitize(dnsRule.MatchPattern)
-				dnsPatternAsRE := matchpattern.ToRegexp(dnsPattern)
-				reStrings = append(reStrings, "("+dnsPatternAsRE+")")
+				dnsPattern := FromFqdn(dnsRule.MatchPattern)
+				dnsPatternAsRE, isAllAllowed := CustomToRegexp(Reverse(dnsPattern))
+				// If the pattern is a wildcard, quit the regex calculation and insert a simple allow
+				// all rule. This will reduce memory usage and improve performance significantly when
+				// a wildcard rule is present together with a set of other rules.
+				// TODO do even need any validation at all here, since we will only ever get valid domains?
+				// eg do only "."
+				if isAllAllowed {
+					reStrings = []string{
+						"|" + allowedDNSCharsREGroup + "+(?:[.]" + allowedDNSCharsREGroup + "+)*",
+					}
+					break
+				}
+				reStrings = append(reStrings, dnsPatternAsRE)
 			}
 		}
-		mp := strings.Join(reStrings, "|")
-		rei, err := re.CompileRegex(mp)
+		// Sort the rules to make faster and smaller regexes, as well as speeding up the compilation.
+		// The impact of sorting is negligible for a low number of rules, and the more rules, the more the
+		// benefit of lower memory usage, faster compilation and faster matching start being noticeable
+		// TODO test this in real life to see for performance improvements
+		sort.Strings(reStrings)
+		// Explicitly add end and beginning of string markers to the beginning and end to avoid having them
+		// defined on a per dns rule basis. Also avoid wrapping each dns rule inside a group since we don't
+		// need to know where a given match occur. All those will also improve the performance of both the
+		// regex compilation and the regex matching
+		// TODO figure out if we can/want use a separate key for the regex cache to avoid the sorting
+		// and extra logic in case its present in the cache
```

Comment:
What do you have in mind as keys to the cache?

Reply:
Lets discuss all this in https://github.com/cilium/cilium/pull/20396 😄